[DO NOT MERGE] new healthcheck

deepthi opened the pull request:
creating a PR to get feedback.

deepthi committed:
healthcheck: rename healthcheck and associated structs to legacy

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: move more code to legacy, and create stubs for new healthcheck

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: remove unused structs, compactify new healthcheck function params

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: move tabletStats into separate file, implement cell map of tabletStatsCache, refactor to encapsulate more behavior that belongs in healthcheck

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: implement topo watch, tablet_stats_cache test, stub new healthcheck tests

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: remove unnecessary flag, fix some TODOs

Signed-off-by: deepthi <deepthi@planetscale.com>

sougou commented on go/vt/discovery/tablet_stats_cache.go:
```diff
@@ -19,132 +19,44 @@ package discovery
-	mu sync.RWMutex
+	mu sync.Mutex
 	// entries maps from keyspace/shard/tabletType to our cache.
 	entries map[string]map[string]map[topodatapb.TabletType]*tabletStatsCacheEntry
```
Change this to a map[string]*legacyTabletStatsCacheEntry where the key is "Keyspace.Shard.TabletType". A further simplification could be to prefix the cell also, unless we need to special-case the current cell.

deepthi replied:
I had thought about key'ing it off querypb.Target, but that includes cell. We have 1 cache per cell so that they can be updated in parallel rather than 1 cache across all cells.

deepthi replied:
This doesn't apply any more since we have deleted tabletStatsCache and replaced with a map-map of *TabletHealth.

sougou commented on go/vt/discovery/healthcheck.go:
```diff
@@ -740,7 +900,7 @@ func (hc *HealthCheckImpl) AddTablet(tablet *topodatapb.Tablet, name string) {
 	key := TabletToMapKey(tablet)
```
This algorithm is not worth it. I think we can just use the tablet alias.

deepthi replied:
+1

sougou commented on go/vt/discovery/healthcheck.go:
```diff
@@ -517,7 +692,8 @@ func (hc *HealthCheckImpl) checkConn(hcc *healthCheckConn, name string) {
 	// Initial notification for downstream about the tablet existence.
-	hc.updateHealth(hcc.tabletStats.Copy(), hcc.conn)
+	// do not copy
+	hc.updateHealth(&hcc.tabletStats, hcc.conn)
```
I don't think `hc` should be updating health. Will be better if stream owned the tabletstats, and kept that up-to-date. `hc` should just consume it.

deepthi replied:
deleted this function

sougou commented on go/vt/discovery/healthcheck.go:
```diff
@@ -136,156 +154,39 @@ func ParseTabletURLTemplateFromFlag() {
+// TabletRecorder is the part of the HealthCheck interface that can
+// add or remove tablets. We define it as a sub-interface here so we
+// can add filters on tablets if needed.
+type TabletRecorder interface {
```
Is there a reason why this is a separate interface instead of being part of Healthcheck?

deepthi replied:
It is still being used by the old healthcheck and offshoots (Filters)

sougou reviewed (commented):
Initial reactions.

deepthi committed:
healthcheck: simplify healthcheck struct by deleting tablet_stats_cache. rename tabletStats to tabletHealth

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: if cells_to_watch is empty, operate on just localCell

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: delete commented code

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: use tabletHealth.mu to serialize access to members. handle tabletType change in healthcheck

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: move shuffleTablets back to gateway

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: Move topoWatch back into TopologyWatcher and use TabletRecorder interface. Move cellsAliases cache into TopologyWatcher

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: remove references to TabletToMapKey in new topo watcher

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: notify buffer when failover ends, rename entries -> healthData

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: rename old tests to legacy

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: move loadTablets from TopologyWatcher to HealthCheckImpl,
implement healthy tablet sorting by replica lag, notify buffer when
master changes.

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: fold healthCheckConn into TabletHealth, fix CacheStatus, get basic unit test working

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: rename TabletHealth->tabletHealthCheck, create a simpler TabletHealth struct that is returned to gateway. Unit tests

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: HealthCheck cache in vtgate debug/status page

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: refactor stream into smaller funcs, use channel for unit tests

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: ScatterConn should return the correct cache status based on gateway_implementation

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: topology_watcher unit tests

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: fix stats and healthy, endtoend test

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: endtoend test for scatter_conn and buffer

Signed-off-by: deepthi <deepthi@planetscale.com>

deepthi committed:
healthcheck: lock hc.mu before accessing members

Signed-off-by: deepthi <deepthi@planetscale.com>
